node/src/main/scala/com/wavesplatform/database/RocksDBWriter.scala

```diff
@@ -484,15 +483,16 @@ class RocksDBWriter(
         expiredKeys ++= updateHistory(rw, Keys.assetDetailsHistory(asset), threshold, Keys.assetDetails(asset))
       }
 
-      val txLeases = for {
-        (_, txInfo) <- snapshot.transactions
-        (id, lease) <- txInfo.snapshot.leaseStates
-      } yield (Some(txInfo.transaction), id, lease)
-      val txLeaseIdSet = txLeases.map(_._2).toSet
-      val allLeases    = txLeases ++ snapshot.leaseStates.collect { case (id, lease) if !txLeaseIdSet.contains(id) => (None, id, lease) }
-      allLeases.foreach { case (txOpt, id, lease) =>
-        val details = lease.toDetails(this, txOpt, leaseDetails(id))
-        rw.put(Keys.leaseDetails(id)(height), Some(Right(details)))
+      for ((id, li) <- snapshot.newLeases) {
+        rw.put(Keys.leaseDetails(id)(height), Some(LeaseDetails(li, snapshot.cancelledLeases.getOrElse(id, LeaseDetails.Status.Active))))
+        expiredKeys ++= updateHistory(rw, Keys.leaseDetailsHistory(id), threshold, Keys.leaseDetails(id))
+      }
+
+      for ((id, status) <- snapshot.cancelledLeases if !snapshot.newLeases.contains(id)) {
```

Comment:
Should we add the `!snapshot.newLeases.contains(id)` filter here? What if there are LeaseTransaction and LeaseCancelTransaction for same leasing in one block?